enrollment-server-onboarding-common/src/main/java/com/wultra/app/onboardingserver/common/database/OnboardingProcessRepository.java

```diff
@@ -40,72 +42,133 @@
      * @return onboarding process IDs
      */
+    @Lock(LockModeType.PESSIMISTIC_WRITE)
     @Query("SELECT p.id FROM OnboardingProcessEntity p " +
             "WHERE p.status = :status " +
             "AND p.timestampCreated < :dateCreatedBefore")
-    List<String> findExpiredProcessIdsByStatusAndCreatedDate(Date dateCreatedBefore, OnboardingStatus status);
-
-    /**
-     * Mark the given onboarding processes as failed.
-     *
-     * @param ids Onboarding process IDs
-     * @param timestampExpired last updated and failed timestamp
-     * @param errorDetail error detail
-     * @param errorOrigin error origin
-     */
-    @Modifying
-    @Query("UPDATE OnboardingProcessEntity p SET " +
-            "p.status = com.wultra.app.enrollmentserver.model.enumeration.OnboardingStatus.FAILED, " +
-            "p.timestampLastUpdated = :timestampExpired, " +
-            "p.timestampFailed = :timestampExpired, " +
-            "p.errorDetail = :errorDetail, " +
-            "p.errorOrigin = :errorOrigin " +
-            "WHERE p.id IN :ids")
-    void terminate(Collection<String> ids, Date timestampExpired, String errorDetail, ErrorOrigin errorOrigin);
+    List<String> findExpiredProcessIdsByStatusWithLock(Date dateCreatedBefore, OnboardingStatus status);
 
     /**
-     * Return onboarding process IDs by the given timestamp. Include only not yet finished entities.
+     * Return onboarding process IDs by the given timestamp. Include only not yet finished entities. Lock these
+     * processes using PESSIMISTIC_WRITE lock until the end of the transaction.
      *
      * @param dateCreatedBefore timestamp created must be before the given value
      * @return onboarding process IDs
      */
+    @Lock(LockModeType.PESSIMISTIC_WRITE)
     @Query("SELECT p.id FROM OnboardingProcessEntity p " +
             "WHERE p.status <> com.wultra.app.enrollmentserver.model.enumeration.OnboardingStatus.FINISHED " +
             "AND p.status <> com.wultra.app.enrollmentserver.model.enumeration.OnboardingStatus.FAILED " +
             "AND p.timestampCreated < :dateCreatedBefore")
-    List<String> findExpiredProcessIdsByCreatedDate(Date dateCreatedBefore);
+    List<String> findExpiredProcessIdsWithLock(Date dateCreatedBefore);
 
     /**
-     * Return onboarding processes to remove activation.
+     * Return onboarding processes to remove activation. Lock these processes using PESSIMISTIC_WRITE lock until
+     * the end of the transaction.
      *
      * @return onboarding processes
      */
+    @Lock(LockModeType.PESSIMISTIC_WRITE)
     @Query("SELECT p FROM OnboardingProcessEntity p " +
             "WHERE p.status = com.wultra.app.enrollmentserver.model.enumeration.OnboardingStatus.FAILED " +
             "AND p.activationId IS NOT NULL " +
             "AND p.activationRemoved = false")
-    Stream<OnboardingProcessEntity> findProcessesToRemoveActivation();
+    Stream<OnboardingProcessEntity> findProcessesToRemoveActivationWithLock();
```

Comment:
🤔 Do you know how does lock work for streams? Are all entries locker or just one at the time?

Reply:
Usage of streams should not matter, the query still translates to `SELECT ... FOR UPDATE` which will lock all entries at once. I will double check this by tracing Hibernate log messages.